core/src/main/java/org/apache/accumulo/core/file/FileOperations.java

```diff
@@ -234,116 +190,86 @@ public FileSystem getFileSystem() {
       return fs;
     }
 
-    protected void setConfiguration(Configuration fsConf) {
+    protected FileStuff configuration(Configuration fsConf) {
       this.fsConf = fsConf;
+      return this;
     }
 
     public Configuration getConfiguration() {
       return fsConf;
     }
 
-    public AccumuloConfiguration getTableConfiguration() {
-      return tableConfiguration;
+    public FileStuff tableConfiguration(AccumuloConfiguration tableConfiguration) {
+      this.tableConfiguration = tableConfiguration;
+      return this;
     }
 
-    /** Check for null parameters. */
-    protected void validate() {
-      Objects.requireNonNull(getFilename());
-      Objects.requireNonNull(getFileSystem());
-      Objects.requireNonNull(getConfiguration());
-      Objects.requireNonNull(getTableConfiguration());
+    public AccumuloConfiguration getTableConfiguration() {
+      return tableConfiguration;
     }
-  }
 
-  /** Builder interface parallel to {@link FileAccessOperation}. */
-  protected interface FileAccessOperationBuilder<SubbuilderType> extends NeedsFile<SubbuilderType>,
-      NeedsFileSystem<SubbuilderType>, NeedsTableConfiguration<SubbuilderType> {
-    // no optional/generic methods.
-  }
-
-  /**
-   * Operation object for performing {@code getFileSize()} operations.
-   */
-  protected class GetFileSizeOperation extends FileAccessOperation<GetFileSizeOperation>
-      implements GetFileSizeOperationBuilder {
-    /** Return the size of the file. */
-    @Override
-    public long execute() throws IOException {
-      validate();
-      return getFileSize(this);
+    public RateLimiter getRateLimiter() {
+      return rateLimiter;
     }
-  }
-
-  /**
-   * Builder interface for {@link GetFileSizeOperation}, allowing execution of {@code getFileSize()}
-   * operations.
-   */
-  public interface GetFileSizeOperationBuilder
-      extends FileAccessOperationBuilder<GetFileSizeOperationBuilder> {
-    /** Return the size of the file. */
-    long execute() throws IOException;
-  }
 
-  /**
-   * Options common to all {@code FileOperation}s which perform reading or writing.
-   */
-  protected static class FileIOOperation<SubclassType extends FileIOOperation<SubclassType>>
-      extends FileAccessOperation<SubclassType> {
-    private RateLimiter rateLimiter;
-
-    /** Specify a rate limiter for this operation. */
-    @SuppressWarnings("unchecked")
-    public SubclassType withRateLimiter(RateLimiter rateLimiter) {
+    public FileStuff rateLimiter(RateLimiter rateLimiter) {
       this.rateLimiter = rateLimiter;
-      return (SubclassType) this;
+      return this;
     }
 
-    public RateLimiter getRateLimiter() {
-      return rateLimiter;
+    /** Check for null parameters. */
+    public void validate() {
+      Objects.requireNonNull(getFilename());
```

Comment:
For the required params (enforced by the fluent API), the check for null can be moved to the method where the param is provided. I suppose the optional params can be checked when they are set, also, but only if we assume the defaults (if the optional methods aren't called at all) are good already (and they should be).

Reply:
Good idea, thanks.